src/apps/racks/services.py

```diff
@@ -0,0 +1,187 @@
+from decimal import Decimal
+from typing import Union
+
+from pydantic import BaseModel
+from sqlalchemy import delete, select, update
+from sqlalchemy.ext.asyncio import AsyncSession
+
+from src.apps.racks.models import Rack
+from src.apps.racks.schemas import (
+    RackBaseOutputSchema,
+    RackInputSchema,
+    RackOutputSchema,
+    RackUpdateSchema,
+)
+from src.apps.sections.models import Section
+from src.apps.sections.services import manage_section_state
+from src.apps.warehouse.models import Warehouse
+from src.apps.warehouse.services import get_all_warehouses, manage_warehouse_state
+from src.core.exceptions import (
+    AlreadyExists,
+    DoesNotExist,
+    IsOccupied,
+    NotEnoughSectionResourcesException,
+    NotEnoughWarehouseResourcesException,
+    RackIsNotEmptyException,
+    ServiceException,
+    TooLittleRackLevelsAmountException,
+    TooLittleRacksAmountException,
+    TooLittleWeightAmountException,
+    WarehouseAlreadyExistsException,
+    WarehouseDoesNotExistException,
+)
+from src.core.pagination.models import PageParams
+from src.core.pagination.schemas import PagedResponseSchema
+from src.core.pagination.services import paginate
+from src.core.utils.orm import if_exists
+
+
+async def create_rack(
+    session: AsyncSession, rack_input: RackInputSchema
+) -> RackOutputSchema:
+    rack_data = rack_input.dict()
+    section_id = rack_data.get("section_id")
+    if not (section_object := await if_exists(Section, "id", section_id, session)):
+        raise DoesNotExist(Section.__name__, "id", section_id)
+
+    if not section_object.available_racks:
+        raise NotEnoughSectionResourcesException(resource="racks")
+
+    if (
+        rack_max_weight := rack_data.get("max_weight")
+    ) > section_object.available_weight:
+        raise NotEnoughSectionResourcesException(resource="available section weight")
+
+    new_rack = Rack(**rack_data)
+    session.add(new_rack)
+    print(rack_max_weight)
+    section = await manage_section_state(
+        section_object,
+        adding_resources_to_section=False,
+        racks_involved=True,
+        weight_involved=True,
+        stock_weight=rack_max_weight,
```

Comment:
i think that there should be monitored the weight that was reserved for the racks
lets say max section weight is 1000, there is 4 racks with max weight of 200 -> 4 x 200 = 800
on every rack is filled with stocks by 50% so -> occupied_weight = 4 x 100 = 400
so - max_section_weight - 1000
reserved_weight -> 800
weight_to_reserve -> 200
occupied_weight -> 400
available_weight -> 600

reserved weight and weight_to_reserve shows to weight that for reserved for the racks in the section, but it does not mean that this weight is occupied by stocks - it just shows that this amount of weight cant be considered as avaialble weight when distributing weight for creating new racks

Reply:
and following this, while managing section state, when weight_involved=True, and stock_object is provided, then we prompt, that some stock has entered the rack of the section, and occupied_weight and available_weight values needs to be changed, but when stock_object is not provided, it will suggest that in this situation, the values of reserved_weight and weight_to_reserve should be modified (of course new parameter should be added with the amount of weight to apply to these modifications)